src/components/vintage/vintage.cairo

```diff
@@ -0,0 +1,233 @@
+            let mut vintage: CarbonVintage = self.Vintage_vintages.read(token_id);
+            vintage.status = new_status;
+            self.Vintage_vintages.write(token_id, vintage);
+        }
+
+        fn set_project_carbon(ref self: ComponentState<TContractState>, new_carbon: u128) {
+            // [Check] Caller is owner
+            self.assert_only_role(OWNER_ROLE);
+
+            // [Check] Project carbon is not 0
+            assert(new_carbon >= 0, 'Project carbon cannot be 0');
+            // [Effect] Update storage
+            let old_carbon = self.Vintage_project_carbon.read();
+            self.Vintage_project_carbon.write(new_carbon);
+            // [Event] Emit event
+            self.emit(ProjectCarbonUpdate { old_carbon, new_carbon, });
+        }
+
+        fn set_vintages(
+            ref self: ComponentState<TContractState>,
+            yearly_absorptions: Span<u128>,
+            start_year: u32
+        ) {
+            // [Check] Caller is owner
+            self.assert_only_role(OWNER_ROLE);
+
+            // [Check] Vintages length is not 0
+            assert(yearly_absorptions.len() > 0, 'Vintages length is 0');
+            let vintages_num = yearly_absorptions.len();
+
+            // [Effect] Update storage
+            self.Vintage_vintages_len.write(vintages_num);
+            let mut index = 0;
+            loop {
+                if index == vintages_num {
+                    break ();
+                }
+                let supply = *yearly_absorptions.at(index);
+
+                let vintage = CarbonVintage {
+                    year: (start_year + index).into(),
+                    supply: supply,
+                    failed: 0,
+                    status: CarbonVintageType::Projected,
+                };
+                self.Vintage_vintages.write(index.into(), vintage);
+                self.emit(VintageUpdate { token_id: index.into(), vintage: vintage, });
+                index += 1;
+            };
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +Drop<TContractState>,
+        +IAccessControl<TContractState>
+    > of InternalTrait<TContractState> {
+        fn initializer(
+            ref self: ComponentState<TContractState>, starting_year: u32, number_of_years: u32
+        ) {
+            // [Storage] Store new vintages
+            let mut index = starting_year;
+            let n = index + number_of_years;
+            loop {
+                if index == n {
+                    break ();
+                }
+                let vintage: CarbonVintage = CarbonVintage {
+                    year: index.into(), supply: 0, failed: 0, status: CarbonVintageType::Projected,
+                };
+                // [Effect] Store values
+                self.Vintage_vintages.write(index.into(), vintage);
+                index += 1;
+            };
+        }
```

Comment:
Do we really need that initializer? If we decide to go on a 2-steps minter, then, we would initialize and set vintages after the end of raise of funds. No need to initialize before the end.